### ISSUE TITLE
E5 model finetuning code

### ISSUE THREAD

**Issue by shamanez** [microsoft/unilm#1005]
It would be amazing to have a fine-tuning script for E5 models. 

Do we need hard negatives when finetuning for downstream tasks?

**Comment by intfloat** [comment]
Sorry for the late response, I somehow missed this issue.

**Q: Do we need hard negatives when finetuning for downstream tasks?**

Yes, E5 fine-tuning needs hard negatives. For NLI, the hard negatives are the contradiction sentences. For MS-MARCO and NQ, the hard negatives are mined with a trained dense retriever.

For fine-tuning script, you can refer to https://github.com/microsoft/unilm/tree/master/simlm . The main difference is on the data side. SimLM is fine-tuned on MS-MARCO only, but E5 is fine-tuned on 3 datasets.

**Comment by intfloat** [comment]
For MSMARCO, we use the same data from https://github.com/microsoft/unilm/tree/master/simlm#download-our-pre-processed-data

For NQ, we did not release the data, but you can use public versions with mined hard negatives.

**Comment by memray** [comment]
@intfloat Thanks! Regarding NQ, the original version is released by Google but it doesn't have positive passages. Both DPR and BEIR have their own NQ training. So I wonder which NQ positives is used in your training?

**Comment by intfloat** [comment]
We use the questions and positive passages from DPR.

**Comment by theejung** [comment]
I have follow-up questions about NLI dataset finetuning.

1. Did you use hard negatives to train cross encoder? And you din't use BM25 retreiver to mine hard negatives for NLI?
2. After fine-tuning cross encoder, did you get inference for the same training data and use it as an input for the knowledge distillation? In there, did you still have 7 negatives?

**Comment by intfloat** [comment]
@theejung Thanks for the questions.

1. Yes, we use hard negatives to train cross encoder, but the hard negatives are contradictory sentences instead of BM25 mined. This strategy is borrowed from [SimCSE paper](https://arxiv.org/abs/2104.08821).
2. Yes, we use the trained cross encoder to label its training data. For NLI, we only have 1 hard negative. During bi-encoder training, we will randomly sample 6 negatives from the entire corpora to make it 7.
3. We randomly choose one sentence from either premise or hypothesis and use "query: " prefix, then assign "passage: " prefix for another sentence.
